File: src/new/ForzaStudio/ForzaStudioCLI/Forza/ForzaTrackSubSection.cs
using System;

namespace Forza
{
    public class ForzaTrackSubSection
    {
        public ForzaTrackSection Parent;

        public EndianStream Stream;

        public string Name;

        public uint Lod;

        public IndexType IndexType;

        public int[] Indices;

        public ForzaVertex[] Vertices;

        public int VertexCount;

        public int FaceCount;

        public Vector3 Offset;

        public Vector2 UVoffset;

        public Vector2 UVtile;

        public ForzaTrackSubSection(ForzaTrackSection parent)
        {
            Parent = parent;
            Stream = parent.Stream;
            Utilities.AssertEquals(Stream.ReadUInt32(), 1u);
            Utilities.AssertEquals(Stream.ReadUInt32(), 2u);
            Name = Stream.ReadASCII(Stream.ReadInt32()).ToLowerInvariant();
            Lod = Stream.ReadUInt32();
            IndexType = (IndexType)Stream.ReadUInt32();
            if (IndexType != IndexType.TriList && IndexType != IndexType.TriStrip)
            {
                throw new Exception("analyze this!");
            }
            uint num = Stream.ReadUInt32();
            Utilities.AssertEquals(Stream.ReadUInt32(), 1u);
            Utilities.AssertEquals(Stream.ReadUInt32(), 0u);
            Utilities.AssertEquals(Stream.ReadUInt32(), 0u);
            Utilities.AssertEquals(Stream.ReadUInt32(), 0u);
            Utilities.AssertEquals(Stream.ReadUInt32(), 0u);
            Utilities.AssertEquals(Stream.ReadSingle(), 1f);
            Utilities.AssertEquals(Stream.ReadSingle(), 1f);
            Utilities.AssertEquals(Stream.ReadSingle(), 1f);
            Utilities.AssertEquals(Stream.ReadSingle(), 1f);
            UVoffset = new Vector2(Stream.ReadSingle(), Stream.ReadSingle());
            UVtile = new Vector2(Stream.ReadSingle(), Stream.ReadSingle());
            Utilities.AssertEquals(Stream.ReadUInt32(), 3u);
            Indices = Utilities.ReadIndices(Stream, Stream.ReadInt32(), Stream.ReadInt32());
            num = Stream.ReadUInt32();
            if (num != 0 && num != 1 && num != 2 && num != 5)
            {
                throw new Exception("analyze this!");
            }
            VertexCount = Utilities.CalculateVertexCount(Indices);
            FaceCount = Utilities.CalculateFaceCount(Indices, IndexType);
        }
    }
}

File: src/py/models/forza_track_subsection.py
import struct
from mathutils import Vector
from models.index_type import IndexType
import utils.mesh_utils

class ForzaTrackSubSection:
    def __init__(self, f):
        
        assert(1 == int.from_bytes(f.read(4), byteorder="big", signed=False))
        assert(2 == int.from_bytes(f.read(4), byteorder="big", signed=False))
        
        # name
        length_bytes = f.read(4)
        length = struct.unpack("<i", length_bytes)[0]
        name_bytes = f.read(length)
        self.name : str = name_bytes.decode("ascii").lower()

        # lod
        self.lod : int = int.from_bytes(f.read(4), byteorder="big", signed=False)

        # index type
        self.index_type : IndexType = IndexType(int.from_bytes(f.read(4), byteorder="big", signed=False))

        # skip and assert
        f.read(4)
        assert(1 == int.from_bytes(f.read(4), byteorder="big", signed=False))
        assert(0 == int.from_bytes(f.read(4), byteorder="big", signed=False))
        assert(0 == int.from_bytes(f.read(4), byteorder="big", signed=False))
        assert(0 == int.from_bytes(f.read(4), byteorder="big", signed=False))
        assert(0 == int.from_bytes(f.read(4), byteorder="big", signed=False))
        assert(1.0 == struct.unpack(">f", f.read(4))[0])
        assert(1.0 == struct.unpack(">f", f.read(4))[0])
        assert(1.0 == struct.unpack(">f", f.read(4))[0])
        assert(1.0 == struct.unpack(">f", f.read(4))[0])

        # uv
        self.uv_offset = Vector((struct.unpack(">f", f.read(4))[0],struct.unpack(">f", f.read(4))[0]))
        self.uv_tile = Vector((struct.unpack(">f", f.read(4))[0],struct.unpack(">f", f.read(4))[0]))

        # assert
        assert(3 == int.from_bytes(f.read(4), byteorder="big", signed=False))

        # indices
        f.read(4)
        self.indices = utils.mesh_utils.read_indices(f, int.from_bytes(f.read(4), byteorder="big", signed=False))

        num = int.from_bytes(f.read(4), byteorder="big", signed=False)
        if num != 0 and num != 1 and num != 2 and num != 5:
            raise RuntimeError("analyze this!")
        
        self.vertex_count = utils.mesh_utils.calculate_vertex_count(self.indices)
        self.face_count = utils.mesh_utils.calculate_face_count(self.indices, IndexType)
